Add Http2 support & envoy

mkocher opened the pull request:
We've updated the release to be able to test an http2 gorouter and updated gostatic to have an envoy in the data path so that it mimics the real world.

We're currently running four benchmarks in the errand - against gorouter h1, gorouter h2, envoy and static. We may want to change the assertions at the end to assert that both h1 and h2 tests are fast enough.

todo: 
- [x] Upload the envoy blob somewhere
- [ ] decide how much of this should run in CI
- [ ] consider deleting the jupyter notebook stuff, it's broken maybe

mkocher committed:
WIP: intial work to support h2

- gostatic supports h2 and h1
- refactored run_perf_test for clarity

ctlong committed:
Uploaded envoy blob and created a bosh package for it

[#177674381]

ctlong committed:
Envoy is going to totally work now

* created job to deploy envoy blob
* copied pipeline manifest to deploy changes and added envoy job to
gostatic instance group

[#177674381](https://www.pivotaltracker.com/story/show/177674381)

belinda-liu committed:
use correct envoy blob and fixed envoy packaging

- took the blob from the diego release by being in the diego-release
directory and running bosh sync-blobs, and copied over the tarball in
the blobs folder

[#177674381](https://www.pivotaltracker.com/story/show/177674381)
[cloudfoundry/routing-release#200](https://github.com/cloudfoundry/routing-release/issues/200)

Co-authored-by: Carson Long <lcarson@vmware.com>
Co-authored-by: Weyman Fung <weymanf@vmware.com>

Gerg committed:
Update hey to point to version wich supports mTLS

Co-authored-by: Matthew Kocher <mkocher@vmware.com>

mkocher committed:
Templatize TLS credentials in envoy, gorouter & perftest

Co-authored-by: Greg Cobb <gcobb@vmware.com>

Gerg committed:
Fix hey packaging to build what's in src

It was accidentally installing from github before this.

Switch spec to include all files in hey package root direction,
specifically to include go.mod file so that it will build correctly.
hashtag gergsub

Co-authored-by: Matthew Kocher <mkocher@vmware.com>

mkocher committed:
Added h2 and envoy runs to the run_perf_test errand

Co-authored-by: Greg Cobb <gcobb@vmware.com>

Gerg committed:
add envoy blob

Co-authored-by: Matthew Kocher <mkocher@vmware.com>

ctlong commented on jobs/performance_tests/templates/run_perf_test.erb:
```diff
@@ -67,29 +90,9 @@ run_perf_tests() {
   NUM_OK=$(grep -E "\[200\].*responses" ${LOG_DIR}/routed_loadtest_${date}.log | awk '{ print $2 }')
 }
 
-emit_datadog_metrics() {
```
Do we really want to delete this function? And if we do, then we should probably remove the call to it on line 104 of this same file.

ameowlia replied:
We don't use datadog anymore that I know of. This is safe to delete.

mkocher replied:
cool, we'll delete *all* the datadog stuff from the release when we clean it up

ameowlia replied:
👍

mkocher committed:
set enable_http2 to true, reduce number of routes

- route_populator/gorouter were struggling to keep up with 100k routes,
  despite neither maxing our their CPU. Scaled back to 50k as that
  wasn't the objective of our test
- set the new gorouter flag to true to enable http2

Co-authored-by: Merric de Launey <mdelauney@vmware.com>

ameowlia commented on jobs/performance_tests/templates/run_perf_test.erb:
```diff
@@ -3,47 +3,70 @@
 
-  echo "Starting direct load test of ${TOTAL_REQUESTS} requests, ${TOTAL_CONCURRENT} concurrent..."
+  # run load test against gorouter using http2
+  echo "Starting gorouter http2 load test of ${TOTAL_REQUESTS} requests, ${TOTAL_CONCURRENT} concurrent..."
```
It looks like the performance tests currently always run h2 tests, but according to this [PR](https://github.com/cloudfoundry/routing-release/pull/205) gorouter will not always have h2 enabled. And currently, since the PR is not merged, no gorouter supports h2 and his test will fail.

❓ Can you make it configurable to run the h2 tests?

ctlong replied:
It might be that we don't look to merge this until after Gorouter does support h2. Are there cases (or CI) where performance tests are run on older versions of Gorouter?

ameowlia replied:
We do not run performance tests against old versions of gorouter.

However, the current PR to routing release and gorouter makes it seem like we are going to have a bosh property to enable h2. If we have a switch there to turn it off, I think we should have a switch here to turn it off too.

ameowlia commented on jobs/performance_tests/templates/run_perf_test.erb:
```diff
@@ -3,47 +3,70 @@
+  echo ""
+
+  # run load test aginst gostatic envoy
+  echo "Starting gostatic envoy test of ${TOTAL_REQUESTS} requests, ${TOTAL_CONCURRENT} concurrent..."
```
❓ Can you explain what this is testing? What is a "gostatic envoy" (this line) vs a "gostatic direct load test" (below)

ctlong replied:
"gostatic envoy" hits the app through envoy using HTTP/1.1 and TLS. "gostatic direct load test" hits the app directly via HTTP/1.1, circumventing envoy.

The purpose IIRC was to determine how much envoy TLS termination was slowing down our requests.

ameowlia replied:
Oh, nice. That makes a lot of sense. Can you add something to that effect in the comments so it is more clear?

ameowlia commented on manifests/bran.yml:
```diff
@@ -0,0 +1,173 @@
+---
+name: perf-test
```
❓ what is bran?

ctlong replied:
@ameowlia it's Mona's favorite kind of muffin, and the arbitrary name we chose for our manifest. We could probably stand to change this.

ameowlia replied:
😹  yeah, please change. Also the correct favorite type of muffin is banana nut muffin 🍌 .

mkocher replied:
Gotta go with @monamohebbi here, Bran muffins are way better than banana nut muffins.

As part of the cleanup phase we want to delete all the other manifests and just leave one modern bosh 2.0 manifest.

ctlong replied:
I would argue that a muffin without any chocolate in it cannot be the best muffin...

ameowlia reviewed (changes requested):
I'm so glad you all are adding envoy into the performance test path.

❓ Can you make sure to edit the readme before this is merged in. Feel free to completely trash the current one, since most of it seems no longer true or relevant.

ameowlia commented:
Do you plan on taking this out of "draft" @mkocher, @ctlong ?

ctlong commented:
@ameowlia we don't have a plan to do so at the moment – I don't remember whether this branch is ready to commit, but I think it probably isn't.
